GVFS/GVFS.Common/Cleanup/GitCleanupQueue.cs

```diff
@@ -0,0 +1,103 @@
+﻿using GVFS.Common.Tracing;
+using System;
+using System.Collections.Concurrent;
+using System.IO;
+using System.Threading;
+
+namespace GVFS.Common.Cleanup
+{
+    public class GitCleanupQueue
+    {
+        private GVFSContext context;
+        private BlockingCollection<GitCleanupStep> queue = new BlockingCollection<GitCleanupStep>();
+        private CancellationTokenSource cancellationToken = new CancellationTokenSource();
+        private GitCleanupStep currentStep;
+
+        public GitCleanupQueue(GVFSContext context)
+        {
+            this.context = context;
+            Thread cleanupWorker = new Thread(() => this.RunQueue());
+            cleanupWorker.Name = "CleanupWorker";
+            cleanupWorker.Start();
+        }
+
+        public void Enqueue(GitCleanupStep step)
+        {
+            this.queue.Add(step);
+        }
+
+        public void Stop()
+        {
+            this.cancellationToken.Cancel();
+            this.currentStep?.Stop();
+        }
+
+        /// <summary>
+        /// This method is public for test purposes only.
+        /// </summary>
+        public bool EnlistmentRootReady()
+        {
+            // If a user locks their drive or disconnects an external drive while the mount process
+            // is running, then it will appear as if the directories below do not exist or throw
+            // a "Device is not ready" error.
+            try
+            {
+                return this.context.FileSystem.DirectoryExists(this.context.Enlistment.EnlistmentRoot)
+                         && this.context.FileSystem.DirectoryExists(this.context.Enlistment.GitObjectsRoot);
+            }
+            catch (IOException)
+            {
+                return false;
+            }
+        }
+
+        private void RunQueue()
+        {
+            while (true)
+            {
+                try
+                {
+                    this.queue.TryTake(out this.currentStep, Timeout.Infinite, this.cancellationToken.Token);
+                }
+                catch (OperationCanceledException)
+                {
+                    // Only gets thrown when stop is requested
+                    return;
+                }
+
+                if (this.EnlistmentRootReady())
+                {
+                    try
+                    {
+                        this.currentStep.Execute();
+                    }
+                    catch (Exception e)
+                    {
+                        this.LogErrorAndExit(
+                            telemetryKey: nameof(GitCleanupQueue),

```

Comment:
`telemetryKey` seems like the wrong word here.  It is more of an message prefix.  We use "Area" in other places in the code so maybe that is the right metadata field to use?